internal/util/streamingutil/service/resolver/resolver_with_discoverer.go

```diff
@@ -0,0 +1,192 @@
+
+	latestStateCond *syncutil.ContextCond
+	latestState     discoverer.VersionedState
+}
+
+// GetLatestState returns the latest state of the resolver.
+func (r *resolverWithDiscoverer) GetLatestState() VersionedState {
+	r.latestStateCond.L.Lock()
+	state := r.latestState
+	r.latestStateCond.L.Unlock()
+	return state
+}
+
+// Watch watch the state change of the resolver.
+func (r *resolverWithDiscoverer) Watch(ctx context.Context, cb func(VersionedState) error) error {
+	state := r.GetLatestState()
+	if err := cb(state); err != nil {
+		return errors.Mark(err, ErrInterrupted)
+	}
+	version := state.Version
+	for {
+		if err := r.watchStateChange(ctx, version); err != nil {
+			return errors.Mark(err, ErrCanceled)
+		}
+		state := r.GetLatestState()
+		if err := cb(state); err != nil {
+			return errors.Mark(err, ErrInterrupted)
+		}
+		version = state.Version
+	}
+}
+
+// Close closes the resolver.
+func (r *resolverWithDiscoverer) Close() {
+	// Cancel underlying task and close the discovery service.
+	r.taskNotifier.Cancel()
+	r.taskNotifier.BlockUntilFinish()
+}
+
+// watchStateChange block util the state is changed.
+func (r *resolverWithDiscoverer) watchStateChange(ctx context.Context, version typeutil.Version) error {
+	r.latestStateCond.L.Lock()
+	for version.EQ(r.latestState.Version) {
+		if err := r.latestStateCond.Wait(ctx); err != nil {
+			return err
+		}
+	}
+	r.latestStateCond.L.Unlock()
+	return nil
+}
+
+// RegisterNewWatcher registers a new grpc resolver.
+// RegisterNewWatcher should always be call before Close.
+func (r *resolverWithDiscoverer) RegisterNewWatcher(grpcResolver *watchBasedGRPCResolver) error {
+	select {
+	case <-r.taskNotifier.Context().Done():
+		return errors.Mark(r.taskNotifier.Context().Err(), ErrCanceled)
+	case r.registerCh <- grpcResolver:
+		return nil
+	}
+}
+
+// doDiscover do the discovery on background.
+func (r *resolverWithDiscoverer) doDiscover() {
+	grpcResolvers := make(map[*watchBasedGRPCResolver]struct{}, 0)
+	defer func() {
+		// Check if all grpc resolver is stopped.
+		for r := range grpcResolvers {
+			if err := lifetime.IsWorking(r.State()); err == nil {
+				r.logger.Warn("resolver is stopped before grpc watcher exist, maybe bug here")
+				break
+			}
+		}
+		r.logger.Info("resolver stopped")
+		r.taskNotifier.Finish(struct{}{})
+	}()
+
+	for {
+		ch := r.asyncDiscover(r.taskNotifier.Context())
+		r.logger.Info("service discover task started, listening...")
+	L:
+		for {
+			select {
+			case watcher := <-r.registerCh:
+				// New grpc resolver registered.
+				// Trigger the latest state to the new grpc resolver.
+				if err := watcher.Update(r.GetLatestState()); err != nil {
+					r.logger.Info("resolver is closed, ignore the new grpc resolver", zap.Error(err))
+				} else {
+					grpcResolvers[watcher] = struct{}{}
+				}
+			case stateWithError := <-ch:
+				if stateWithError.err != nil {
+					if r.taskNotifier.Context().Err() != nil {
+						// resolver stopped.
```

Comment:
Log stateWithError.err before returning.

Reply:
It's the normal path when resolver is closed.
so "resolver stopped" has been logged at defer function.